lib/AST/ModuleNameLookup.cpp

```diff
@@ -254,194 +114,114 @@ class LookupVisibleDecls : public ModuleNameLookup<LookupVisibleDecls> {
-  // imports if needed.
-  const bool canReturnEarly = recordImportDecls(decls, scopedValues, overloads);
-  if (!canReturnEarly)
-    (void)recordImportDecls(decls, unscopedValues, overloads);
-}
+      // Skip the special import set module; we've already visited it.
+      if (import.second == headerImportModule)
+        continue;
 
-template <typename LookupStrategy>
-ArrayRef<ValueDecl *> ModuleNameLookup<LookupStrategy>::lookupInModuleUncached(
-    SmallVectorImpl<ValueDecl *> &decls,
-    ModuleDecl *module, ModuleDecl::AccessPathTy accessPath,
-    const DeclContext *moduleScopeContext,
-    ArrayRef<ModuleDecl::ImportedModule> extraImports) {
-
-  // Do the lookup.
-  SmallVector<ValueDecl *, 4> localDecls;
-  getDerived()->doLocalLookup(module, accessPath, localDecls);
-  if (respectAccessControl) {
-    llvm::erase_if(localDecls, [=](ValueDecl *VD) {
-      return !VD->isAccessibleFrom(moduleScopeContext);
-    });
-  }
+      visitImport(import, moduleScopeContext);
+    }
 
-  // Record the decls by overload signature.
-  const size_t initialCount = decls.size();
-  typename LookupStrategy::OverloadSetTy overloads;
-  const bool canReturnEarly = recordImportDecls(decls, localDecls, overloads);
+    for (auto import : imports.getTransitiveImports()) {
+      // Skip the special import set module; we've already visited it.
+      if (import.second == headerImportModule)
+        continue;
 
-  // If needed, search for decls in re-exported modules as well.
-  if (!canReturnEarly) {
-    SmallVector<ModuleDecl::ImportedModule, 8> reexports;
-    module->getImportedModulesForLookup(reexports);
-    assert(llvm::none_of(reexports,
-                         [module](ModuleDecl::ImportedModule import) -> bool {
-                           return import.second == nullptr || import.second == module;
-                         }));
-    reexports.append(extraImports.begin(), extraImports.end());
-
-    // Special treatment based on the use site only applies to immediate
-    // imports of the top-level module.
-    // FIXME: It ought to apply to anything re-exported by those immediate
-    // imports as well, since re-exports are supposed to be treated like part of
-    // the module they're re-exported from.
-    const DeclContext *moduleScopeContextForReexports = moduleScopeContext;
-    if (moduleScopeContext && moduleScopeContext->getParentModule() != module)
-      moduleScopeContextForReexports = nullptr;
-
-    collectLookupResultsFromImports(decls, reexports, accessPath,
-                                    moduleScopeContextForReexports, overloads);
+      visitImport(import, nullptr);
+    }
   }
 
+  // Nothing more to do if we don't have ambiguous results.
+  if (decls.size() - initialCount <= 1)
+    return;
+
   // Remove duplicated declarations, which can happen when the same module is
-  // imported indirectly through two intermediate modules.
+  // imported with multiple access paths.
```

Comment:
I don't think that's correct. Then you wouldn't get duplicates.

Reply:
What if you have "import Foo; import Foo.Bar"?

Reply:
Oh right.